src/utils/helpers/structure/generateProgress.ts

```diff
@@ -0,0 +1,82 @@
+import { TemplateElementCategory } from '../../generated/graphql'
+import { ElementStateNEW, FullStructure, PageNEW, Progress } from '../../types'
+
+const initialProgress = {
+  doneNonRequired: 0,
+  doneRequired: 0,
+  completed: false,
+  totalNonRequired: 0,
+  totalRequired: 0,
+  totalSum: 0,
+  valid: true,
+}
+
+const calculateCompleted = (totalSum: number, doneRequired: number, doneNonRequired: number) => {
+  const totalDone = doneRequired + doneNonRequired
+  return totalSum === totalDone
+}
+
+const getSectionProgress = (pages: PageNEW[]): Progress => {
+  return pages.reduce((sectionProgress: Progress, { progress }) => {
+    if (!progress) return sectionProgress
+    let {
+      doneNonRequired,
+      doneRequired,
+      completed,
+      totalNonRequired,
+      totalRequired,
+      totalSum,
+      valid,
+    } = sectionProgress
+    doneNonRequired += progress.doneNonRequired || 0
+    totalNonRequired += progress.totalNonRequired || 0
+    doneRequired += progress.doneRequired || 0
+    totalRequired += progress.totalRequired || 0
+    totalSum = totalRequired + doneNonRequired
+    completed = calculateCompleted(totalSum, doneRequired, doneNonRequired)
+    if (!progress.valid) valid = false
+    return {
+      doneNonRequired,
+      doneRequired,
+      completed,
+      totalNonRequired,
+      totalRequired,
+      totalSum,
+      valid,
+    }
+  }, initialProgress)
+}
+
+export const generateResponsesProgress = (structure: FullStructure) => {
+  Object.values(structure.sections).forEach((section) => {
+    Object.values(section.pages).forEach((page) => {
+      const progress: Progress = initialProgress
+      page.state
+        .filter(({ element }) => {
+          const { category, isVisible, isEditable } = element as ElementStateNEW
+          return isVisible && isEditable && category === TemplateElementCategory.Question
+        })
+        .forEach(({ element, response }) => {
+          const { isRequired } = element as ElementStateNEW
+          if (isRequired) progress.totalRequired++
+          else progress.totalNonRequired++
+          if (response?.text) {
+            if (isRequired) progress.doneRequired++
+            else progress.doneNonRequired++
+          }
+          if (response !== null && !response?.isValid) progress.valid = false
+          progress.totalSum = progress.totalRequired + progress.doneNonRequired
+          progress.completed = calculateCompleted(
+            progress.totalSum,
+            progress.doneRequired,
+            progress.doneNonRequired
+          )
+        })
+      page.progress = progress
+    })
+    section.progress = getSectionProgress(Object.values(section.pages))
+    section.invalidPage =
```

Comment:
this is not strict right ?

maybe good idea to also do strict here (in another parameter ?

Reply:
I don't know if it would be needed. Do you think it would? When we are running the validation on `strict` mode the first page considered invalid is the first one without the `completed` field as true. Checking for incomplete as well and not only invalid

Reply:
> I don't know if it would be needed. Do you think it would? When we are running the validation on strict mode the first page considered invalid is the first one without the completed field as true. Checking for incomplete as well and not only invalid

Yeah, never mind, it would be super easy to calculate when we need it